### DISCUSSION TITLE
CSV Import / Export: discussion without coding request.

### DISCUSSION THREAD

**Discussion by meteos77** [textbrowser/biblioteq#202] [General]
Report on CSV IMPORT/EXPORT

the complete CSV import has 25 columns according to the template "BiblioteQ Books (All Fields)
3 more fields are required than in the export : description ; marc_tags ; url
the export is not in the same order as the import

the export of the category book has 23 columns
2 more fields : Availability and Total Reserved and so 3 less fields : description ; marc_tags ; url

I tested the fact to have a quantity different from 1 for example 2
In BiblioteQ, the librarian modifies the 2nd copy with the adjustment of the fields Originality; Condition or Status.
When exporting this data is lost, BiblioteQ indicates 2 in quantity but nothing on the characteristics of the different copies.

1 - the CSV import is different from the CSV export (different in number of columns and in field selection)
2 - if some fields are processed : description, marc_tags, url : not present in the CSV export
3 - if copy creation : Originality, Condition and Status are lost because not present in the CSV export

even without processing the import or export flow, the result Export(Import(CSV)) = CSV is false

**Comment by textbrowser** [comment]
Alright. So, I would like you to write a specification file describing the exact behavior of the import. It should include exception handling, duplicate data, missing data, reports, etc. It should describe any present and future scenario and be able to handle any data. It should be perfect and simply work under any circumstance.

This is the task I'm asked of in a few or not-so few words.

I was describing the ideal of the two processes. They are truly inverses of one another. URLs, descriptions, and some other fields are not present in the table. Hence, they do not exist in the export. The export is an export of widgets and not a direct copy of the contents of the respective database. When I said they are inverses, I was referring to conceptual entities.

The patron export is an export of actual database content.

**Comment by textbrowser** [comment]
Can you provide me with a count of the items which cause problems in the import? Please include uniqueness problems and problems of copies. Whatever you consider a problem. Then, break those down into percentages with respect to successes. We can then consider this discussion in percentages and decide whether or not we need to perfect a tool which takes a lot of my time.

**Reply by meteos77** [reply]
I can't predict the future

**Comment by meteos77** [comment]
First of all I think BiblioteQ is fantastic !

If it's a mountain, and the development of the CSV import/export function is not interesting for you; don't do anything on this functionality, it's not a problem.

you have a technical vision that I don't have.
For me: "URLs, descriptions, and some other fields are not present in the table" has no concrete reality.

From my current experience, I can see that to switch from one software to another, the csv format is the most common.
If we work for 10 years with BiblioteQ and at the time of the future switch to another software, data are lost because they are not exported, I prefer to be aware of this from the first use of BiblioteQ. This seems normal to me as a user.
We will take precautions to avoid non-exportable fields.

**Reply by meteos77** [reply]
imagine during ten years, I position the identifier ark:// of the books of the Bnf in the url field
this url field is convenient but not exportable but becomes the reference instead of isbn.

source : https://n2t.net/e/ark_ids.html

**Reply by textbrowser** [reply]
> First of all I think BiblioteQ is fantastic !
> 
> If it's a mountain, and the development of the CSV import/export function is not interesting for you; don't do anything on this functionality, it's not a problem.
> 
> you have a technical vision that I don't have. For me: "URLs, descriptions, and some other fields are not present in the table" has no concrete reality.
> 
> From my current experience, I can see that to switch from one software to another, the csv format is the most common. If we work for 10 years with BiblioteQ and at the time of the future switch to another software, data are lost because they are not exported, I prefer to be aware of this from the first use of BiblioteQ. This seems normal to me as a user. We will take precautions to avoid non-exportable fields.

Take a look, it's in a book. Imagine Balzac writing a story and being pressed on the difficulty of translating the story into English. Hey chap, that French doesn't translate well. Best avoid the creative spirit there. Whatever the future holds with respect to BQ is irrelevant to BQ. It's a creative, open process. If it's dropped, the burden lies on the new experience.

**Reply by meteos77** [reply]
What the future holds for BQ is not irrelevant to meteos77
the experience can also be used to avoid making the same mistakes.

if 20 years ago the people at the library had entered isbn-10 or isbn13 in my current software ... 8000 books would not have to be reanalyzed

**Comment by textbrowser** [comment]
Anyway, this discussion isn't making me happy. I'll introduce the exceptions for copies that we agreed upon and move onto other things. I'm not Microsoft to promise indefinite support. :)

**Reply by meteos77** [reply]
If you were MS, I wouldn't be interested in your software :-)
If it doesn't make you happy, don't make exceptions; that's not the point; you have to take care of yourself.

**Reply by meteos77** [reply]
as I often remind you for my requests  I don't really know if it's mountains or small hills, just sort it out and have fun

**Comment by meteos77** [comment]
the solution exists with SQL
select * from book then export csv view -> I have all the fields (32 in all)

Question: remain the copies to not lose their barcode and their states is it possible with a SQL query or is it a moon shot?

**Comment by textbrowser** [comment]
Did you compute the percentages? How did BQ help you with the import? 50%? 90%? 99%?

**Reply by meteos77** [reply]
I can't because since BQ doesn't do any specific processing of the csv, I can't calculate its failure rate.
If I give him what he expects I have 100%. Now I understand that the input csv is a simple import without data modification.

If I give him in raw my csv output from my old software to the import without modification, I am at 0% of import.

I have created a bash script to avoid losing as much information as possible from the old database. it is not yet 100% perfected but I have to put it back.

currently 
- I have to invent isbn
- I have to concatenate information to put them in the "keyword field
ex : BEGINNING : Date of purchase = 38035 ; Origin = Gift ; Type of Document = ; Format = Pocket ; Audience = Unknown target audience ; END
it is far from being perfect but the main thing is there, I have my 7500 books in the library.

- concerning the arrivals of the Library Department (unimarc file), I pass by catmandu to translate the unimarc in csv then import in BQ.

I have not yet tested perl (catmandu) on the library computer; I hope it will work without problems. I want other library volunteers to be able to do the unimarc import if I'm away.

**Reply by textbrowser** [reply]
This is perplexing. I'm baffled at the amount of work I put into the import. In 45 or so minutes, I was able to catalog 60 books using a barcode scanner along with some manual edits. In 7.5 hours, that would be 600 books. For 10,000 items, that's 125 hours or about 16 days if 8 hours per day. Anyway.

**Reply by textbrowser** [reply]
If I were given this task, I would determine which items have ISBNs and input them using the standard approach. A batch mode will not be ready.

**Reply by meteos77** [reply]
First of all congratulations and thanks for the great improvement of the CSV import tool that I have experienced since November.
Your quantity and quality of work is really felt on this feature.

- With your improvements and my script I switch my actuallalized database of the day to BiblioteQ in 2 minutes. (extraordinary)
of course improvements are always possible but I am very satisfied.

- your efforts on the CSV import you must see it first as a great service to help some future users (and thus time spent to help them).

metaphor: The front of a restaurant. For the chef ... the front bof I prefer to go into my kitchen but for the customer; he enters or not.

- I don't dispute your calculation but it's ideal :-) I had made a survey in november (2x6hours) to see if it wasn't easier to take everything from zero.

    - 1 All books not isbn ( no scan possible )
    - 2 Books have the isbn but not in barcode form and on the first inside page (no scanning possible)
    - 3 All books are not on a table - time to move
    - 4 We do not have internet in the library
    - 5 A fixed computer with the books beyond 10 meters ...

For our case, I think that the CSV import is the most efficient for our constraints.

**Reply by meteos77** [reply]
the major drawback of this method is that we lose information that can be useful
access number and / or internal barcode
Date of purchase
Origin
Public

**Reply by meteos77** [reply]
**The Import CSV feature does its job very well  a big thank you for this development effort**

**Reply by meteos77** [reply]
feedback :  we started to correct the current base result: **200 books in 3 hours**

still a lot of work to do but it's going faster than I thought
the work -> 
-sorting of the books to be treated (not those on loan from the central library)
- handling
- scan of those with ISBN13 to complete the database
- printing of the ISBN13 of the ones with the ISBN10 inside without barcode (the longest ones) then scanning.
- handling
- sorting

30 books printed before 1977 so no isbn10 for the moment they are put aside.

**CSV import is a wonderful feature to switch from an old software to BilioteQ.**

**Comment by textbrowser** [comment]
https://en.wikipedia.org/wiki/Google_Books

So even Google had lots of technical problems. And they're like a huge multi-billion corp. It wasn't just legal problems.

**Reply by meteos77** [reply]
I will not cry for them :-)

**Comment by meteos77** [comment]
**mea cupla**: I think I misunderstood the CSV export feature from the beginning.
It is well titled "Export Table **View** as CSV" so it is normal that I do not have what I expected, the **View** makes all the difference

for me in the future a function Export Table Books as CSV would be a plus.
I take my current example: I start from a csv export to import it into BQ. 
If my current software did not provide this export how would I do it?

the SQL query :
SELECT accession_number, author, binding_type, callnumber, category, condition, description, deweynumber, edition, id, isbn13, keyword, language, lccontrolnumber, location, marc_tags, monetary_units, originality, pdate, place, price, publisher, quantity, title, url FROM BOOK
checks well  Export(Import(CSV)) = CSV

I still have this question about the copy export which should appear as rows in the csv file because CSV does not take care of ISBN duplicates.
